designcompose/src/main/java/com/android/designcompose/squoosh/DESIGN.md

```diff
@@ -0,0 +1,92 @@
+   into the "to" tree.
+ - `Modifier.squooshRender` knows how to render an entire `SquooshResolvedNode`--typically
+   corresponding to the entire Composable.
+
+## Typical Execution
+Whenever squoosh needs to render, it will:
+
+1. Build a new `SquooshResolvedNode` tree with `resolveVariantsRecursively`
+2. Create a request to JNI layout using `updateLayoutTree`
+3. Submit the request to JNI layout using `SquooshLayout.doLayout`
+4. Put the results of the layout into the `SquooshResolvedNode` tree so that rendering is easier
+   using `populateComputedLayout`
+5. If there are transitions, then generate a new `InteractionState` (and soon `CustomizationContext`)
+   with the transition values applied (these state objects start out without the action of the
+   transition applied, then we ask them to make a copy with all active transitions applied -- so it
+   is the state of the world once all transitions are complete). We then run through steps 1-4 again
+   with the updated context.
+    * This gives us a `SquooshResolvedNode` tree of the world after all transitions have ended.
+    * We use `newSquooshAnimate` to create `SquooshAnimationControl` objects that control progress
+      for each transition. `newSquooshAnimate` identifies similar nodes and different nodes and
+      creates objects that will animate the appropriate property for each. Some nodes from the source
+      tree get moved to the destination tree so that they can be faded out.
+6. If there are transitions, we launch an effect to run an animation clock and update the state of
+   each transition's animation. This is how we can seemingly run many different animations at once
+   -- they all just show different positions between the two trees, and can't be nested.
+7. We emit a `Layout` Composable that occupies the appropriate space and has a `squooshRender`
+   Modifier attached.
+
+## Nested Composables and input events
+Squoosh still uses Compose extensively! Child Composables (including `AndroidView`) can be added,
+and this is supported by having `resolveVariantsRecursively` generate a list of child Composables
+that should be emitted (including externally supplied child Composables, and anything in the
+DesignCompose tree that needs input).
+
+Inside of the `Layout` Composable in `SquooshRoot`, we nest any Child Composables from our list.
+
+Painting just one child Composable is tricky. Compose only lets us paint *all* child Composables in
+one go. So we have a horrible hack where a boxed object is shared between the child Composables and
+the `squooshRender` Modifier. When `squooshRender` wants to paint a particular child Composable it
+updates the boxed object (a `SquooshRenderSelector`) to have the id of the node that should be
+rendered. All of the child Composables have a Modifier that customizes drawing to completely skip
+drawing unless the id in the `SquooshRenderSelector` matches. Yuck! But good enough...
```

Comment:
I don't really follow this. In what scenarios do you paint just one child composable?

Reply:
We can only paint all children at once (via `drawContent`), but `squooshRender` wants to render all of the content in one go -- including the stuff before an external child, between two external children, and after the last external child.

I added this thing as a mechanism so that `squooshRender` could call `drawContent` any time it got to a node with an external Composable, but only one of the external Composables would actually render. So as `squooshRender` iterates the tree it will call `drawContent` N times, but each individual external Composable will only get painted once.

I'd really like to be able to have a `drawChild(idx)` method in the `drawWithContent` modifier, but I couldn't figure out how to achieve this, because Compose works quite hard to prevent any modifier from enumerating its children. Perhaps there's a way to do this with the `Modifier.Node` API...

Reply:
I improved the doc in this PR, but I also put it in [Docs](https://docs.google.com/document/d/1tCNgkqCUrb-D1w13nBueeEmSB-ElroriLTyribkuDiE/edit?tab=t.0) so we can iterate on it a bit more easily.